flocker/apiclient/_client.py

```diff
@@ -522,25 +603,32 @@ def create_dataset(self, primary, maximum_size=None, dataset_id=None,
             dataset[u"maximum_size"] = maximum_size
         request = self._request(b"POST", b"/configuration/datasets",
                                 dataset, {CREATED},
-                                {CONFLICT: DatasetAlreadyExists})
+                                {CONFLICT: DatasetAlreadyExists,
+                                 PRECONDITION_FAILED: ConfigurationChanged},
+                                configuration_tag=configuration_tag)
         request.addCallback(self._parse_configuration_dataset)
         return request
 
-    def move_dataset(self, primary, dataset_id):
+    def move_dataset(self, primary, dataset_id, configuration_tag=None):
         request = self._request(
             b"POST", b"/configuration/datasets/%s" % (dataset_id,),
-            {u"primary": unicode(primary)}, {OK})
+            {u"primary": unicode(primary)}, {OK},
+            {PRECONDITION_FAILED: ConfigurationChanged},
+            configuration_tag=configuration_tag)
         request.addCallback(self._parse_configuration_dataset)
         return request
 
     def list_datasets_configuration(self):
-        request = self._request(b"GET", b"/configuration/datasets", None, {OK})
+        request = self._request(b"GET", b"/configuration/datasets", None, {OK},
+                                with_headers=True)
         request.addCallback(
-            lambda results:
-            [
-                self._parse_configuration_dataset(d)
-                for d in results if not d['deleted']
-            ]
+            lambda (results, headers):
+            DatasetsConfiguration(
+                tag=headers.getRawHeaders('X-Configuration-Tag')[0],
```

Comment:
:grey_question: IIUC, using this client with a version of flocker before #2225 is merged would result in a `list index out of range` error instead of some nice version mismatch error, or just handling the cross version incompatability.

I guess given that the flocker client is not its own package people should not be expecting any consumer of this to be cross version compatible. I guess if you think the odds of someone using this client with a pre #2225 server is low fine as is. Otherwise it'd be nice to at least have a nicer error here.